plugins/filter_aws/aws.h

```diff
@@ -40,6 +40,35 @@
 #define FLB_FILTER_AWS_HOSTNAME_KEY                       "hostname"
 #define FLB_FILTER_AWS_HOSTNAME_KEY_LEN                   8
 
+/* defines returned value for cases when configuration is invalid and program should exit */
+#define FLB_FILTER_AWS_CONFIGURATION_ERROR -100
+
+/* defines a group of information, is used as an index in the ctx->metadata_groups
+ * to be specific, it is an index at array flb_filter_aws->metadata_groups */
+#define FLB_FILTER_AWS_METADATA_GROUP_BASE 0
+#define FLB_FILTER_AWS_METADATA_GROUP_TAGS 1
+
+#define FLB_FILTER_AWS_METADATA_GROUP_LEN 2 /* used to define required memory */
+
```

Comment:
In the future, would you plan to support user defined metadata groups? How would that work?

While I think this code is cool and fine, right now, I don't think there's a difference between the code you have and something simple like this:

on the context:
```
struct flb_filter_aws_metadata_group tag_group
struct flb_filter_aws_metadata_group meta_group
```

and then the code just references them like `ctx=>tag_group`. 

Your current model where theres an array and an index in the array works fine. In some ways if you think about it its equivalent to what I show above, since indexing memory in an array is similar to referencing a variable name in a structure which is really just its memory position.

Reply:
> In the future, would you plan to support user defined metadata groups?

I wouldn't want to allow users to define metadata groups. Metadata groups are (and will be) hardcoded.

In the future, we can define a separate metadata group for each piece of information. Each of {`az`, `ec2_instance_id`, `ec2_instance_type`, `private_ip`, `ami_id`, `account_id`, `hostname`, `vpc_id`, `tags`} would be a separate metadata group. This allows to separate the error handling and retries for each of the groups. After all, each of these HTTP requests can fail. Thanks to this approach, since we want to inject as much information as we can, Filter AWS would inject all pieces of information it was able to fetch. (I didn't check the code right now, but I believe in the current implementation, if one of the requests fails during fetching the current `base` metadata group, then we won't inject any `base` information into the logs.)

> In some ways if you think about it its equivalent to what I show above

Yeah, I think it would be the same. As an explanation, my idea was that we can implement the functions for retries in a generic way by passing an index of the array as the argument. However, we can achieve the same thing by simply passing the `&flb_filter_aws_metadata_group`, so there really is no difference in practice.

@PettitWesley I will change the code to use your proposal:
```
struct flb_filter_aws_metadata_group tag_group
struct flb_filter_aws_metadata_group meta_group
```